File: Finished/Clone Graph.java
import java.util.ArrayList;
import java.util.List;
import java.util.*;

// Definition for a Node.
class Node {
  public int val;
  public List<Node> neighbors;

  public Node() {
    val = 0;
    neighbors = new ArrayList<Node>();
  }

  public Node(int _val) {
    val = _val;
    neighbors = new ArrayList<Node>();
  }

  public Node(int _val, ArrayList<Node> _neighbors) {
    val = _val;
    neighbors = _neighbors;
  }
}

class Solution {
  public void dfs(Node node, Node dup, Node[] visited) {
    visited[node.val] = dup;

    for (Node n : node.neighbors) {
      if (visited[n.val] == null) {
        Node dupNode = new Node(n.val);
        dup.neighbors.add(dupNode);
        dfs(n, dupNode, visited);
      } else {
        dup.neighbors.add(visited[n.val]);
      }
    }
  }

  public Node cloneGraph(Node node) {
    if (node == null)
      return null;
    Node dup = new Node(node.val);
    Node[] visited = new Node[101];
    Arrays.fill(visited, null);
    dfs(node, dup, visited);
    return dup;
  }
}

File: finished/clone_graph.py
from collections import deque
from typing import Optional

# https://leetcode.com/problems/clone-graph/

"""
# Definition for a Node.
"""


class Node:
    def __init__(self, val=0, neighbors=None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []


class Solution:
    def cloneGraph(self, node: Optional["Node"]) -> Optional["Node"]:
        visited_original = set()
        cloned = {}  # Dict[<original_node>, <cloned_node>]

        q = deque([node])
        while q:
            n = q.popleft()

            if not n or n in visited_original:
                continue
            visited_original.add(n)

            cln = Node(n.val)
            cloned[n] = cln

            for nb in n.neighbors:
                q.append(nb)

        for org in visited_original:
            cln = cloned[org]

            for org_nb in org.neighbors:
                cln.neighbors.append(cloned[org_nb])

        return cloned[node] if node else None
